### PR TITLE
Major refactoring for reusability

### PR DESCRIPTION
**Pull request by julianbei**
First things first,
thanks for the effort and for sharing this module. This was the only module I found which was more or less sufficient for most of my cases. From my perspective it is worth developing this module because it closes a gap in the ecosystem between angular and bootstrap which is a common setup.
What is completely missing is a proper documentation. Thats somehow the reason for my pull request.
I tried to figure out how your module is supposed to be used. While doing so, i stumbled upon a couple of things which might make sense in other setups, but in my setups they just don´t work.
So I decided to help you to improve this wonderful module.

These are my changes and suggestions:
- Bootstrap is not really a dependency of this package.
  There are several alternatives out there so the bootstrap dependency fixes the usage of this package to the "main bootstrap package". e.g.: [angular-ui-bootstrap](https://github.com/angular-ui/bootstrap)
- The code base was not readable for me. So I encapsulated it a bit and replaced all double tab indentions with single tab indentions.
- The module relied on the build in email validation which is crap in most of the cases. Whats really missing is a "strictemail" validation, so I have added it.
- The attribute for a custom message `title-{{validationName}}` was not meaningful to me because it is not a title, it is more like a notification.
- For my cases it is not sufficient to execute the errorMessage generator on submit only.
  For this case, I have the backend validations which are displayed to the frontend if they fail.
  Frontend validations are for preventing the user to submit an invalid form. So in my cases the submitbutton is disabled if the form is invalid. 

``` html
   <button 
     ng-disabled="form.$invalid" 
     type="submit" 
     class="btn btn-primary pull-right">
     sign up
   </button>
```
- Your comments in the code are very clear and explaining what happens next. - I know this is a question of taste - No one reads comments in code, especially if they are that extensively used. Instead they decrease the readability of your code. A proper Documentation would be worth the effort.
  The comments were mostly TLDR;

So said, the errorMessage generator now executes when:
- the input is touched,
- the inputs validity changes,
- the form gets submitted and is invalid

Lastly, the bower file had some issues because it was loading the from file.
So automatic injection failed.

please review and comment.

What is missing from my perspective:
- [ ] Readme with examples
- [ ] Tests, Tests, Tests (I have not yet added some, but first things first...)
- [ ] configuration options
- [ ] more customisation options

### CODE REVIEW

**Comment by sagrawal31** [comment]
Hi @julianbei Thanks for taking the time to contribute and refactor things. I'm really glad that you fount it helpful. I really appreciate this. I learned new things from your changes like that `main` key in `bower.json`. I never get a chance to read about the `main` key.

All the things you mentioned about email validation and `title-{{validationName}}` key are nice improvements. I've been using 4 spaces for a very long time and that is why I used to prefer it 😄 

Few of the things you mentioned like more examples and documentation are in my pipeline and those have moved higher in priority now since you found it helpful.

I've taken a quick look at your changes and I'll review it closely asap after pulling your changes since changing the indentation made it bit harder to review.

Overall, love your contribution & changes. Thanks

**Comment by sagrawal31** [comment]
@julianbei I looked at your changes and they looks good except the indentation thing I mentioned 😄 I'll merge this code and some configuration options with some examples.

File: src/services/validation.service.js [hunk 1]
```diff
@@ -1,93 +1,112 @@
...
-            return false;
-        },
-        getDefaultMessage: function (key) {
-            return messages[key];
-        }
-    };
-});
\ No newline at end of file
+    getMeta: function (){
+      return ['matchName'];
```
Comment (sagrawal31) [inline src/services/validation.service.js:75]: @julianbei What is the use of this?

Comment (julianbei) [inline src/services/validation.service.js:75]: If you have a password matching like this:

``` html
<form name="form" ng-submit="su.signUp()" novalidate>

          <div class="form-group">
            <input type="email"
              class="form-control"
              name="email"
              placeholder="Email"
              ng-model="su.user.email"

              bs-validation
              strictemail="true"
              required="true"
              response-notification="{{su.error.email}}">

          </div>

          <div class="form-group">
            <input type="password"
              class="form-control"
              name="password"
              placeholder="Password"
              ng-model="su.user.password"

              bs-validation
              ng-minlength="3"
              required="true"
              response-notification="{{su.error.password}}">

          </div>

          <div class="form-group">
            <input type="password"
              class="form-control"
              name="password_confirmation"
              placeholder="Password Confirmation"
              ng-model="su.user.password_confirmation"

              bs-validation
              equalTo="{{su.user.password}}"
              matchName="password"
              required="true"
              response-notification="{{su.error.password_confirmation}}">
          </div>

          <p>
            already an admin? <a ui-sref="main.auth.signin">sign in!</a>
          </p>
          <div class="control-group">
            <div class="controls">
              <button ng-disabled="form.$invalid" type="submit" class="btn btn-primary pull-right">sign up</button>
            </div>
          </div>
        </form>
```

you don´t want to have `must match {{whats in the password input}}` but rather `must match password`

so you need something which is the meta information for 'password' in this case.

Comment (julianbei) [inline src/services/validation.service.js:75]: actually if you get a 422 from the Server with some meta information like an error message for each field.
You can set the validity of each form element with:

``` javascript
let f = this.$scope.form;
    [ f.email, f.password, f.password_confirmation].forEach(
      item => item.$setValidity('response', false)
    );
```

and show the error with: `response-notification="{{su.error.email}}"` 

thats awesome.
but you have to validate the field 'on-blur' manually in the controller. But thats acceptable

Comment (sagrawal31) [inline src/services/validation.service.js:75]: Oh! right. 👍

**Comment by julianbei** [comment]
Thanks

**Pull request merged**